File: backend-java/src/main/java/com/pdfdiff/service/DiffEngine.java
package com.pdfdiff.service;

import com.pdfdiff.domain.LineUnit;
import com.pdfdiff.domain.RawChange;
import com.pdfdiff.domain.Segment;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

@Service
public class DiffEngine {

    public List<RawChange> diffLines(List<LineUnit> templateLines, List<LineUnit> contractLines) {
        List<Segment> tplSegments = splitSegments(templateLines);
        List<Segment> conSegments = splitSegments(contractLines);

        List<List<LineUnit>> tplContent = tplSegments.stream()
                .filter(s -> "content".equals(s.kind()))
                .map(Segment::lines)
                .toList();
        List<List<LineUnit>> conContent = conSegments.stream()
                .filter(s -> "content".equals(s.kind()))
                .map(Segment::lines)
                .toList();
        List<List<LineUnit>> tplEmpty = tplSegments.stream()
                .filter(s -> "empty".equals(s.kind()))
                .map(Segment::lines)
                .toList();
        List<List<LineUnit>> conEmpty = conSegments.stream()
                .filter(s -> "empty".equals(s.kind()))
                .map(Segment::lines)
                .toList();

        if (tplContent.size() == conContent.size() && tplContent.size() >= 2) {
            List<RawChange> changes = new ArrayList<>();
            for (int i = 0; i < tplContent.size(); i++) {
                changes.addAll(diffContentSection(tplContent.get(i), conContent.get(i)));
            }
            int emptyPairs = Math.min(tplEmpty.size(), conEmpty.size());
            for (int i = 0; i < emptyPairs; i++) {
                changes.addAll(diffEmptyRun(tplEmpty.get(i), conEmpty.get(i)));
            }
            for (int i = conEmpty.size(); i < tplEmpty.size(); i++) {
                changes.addAll(deleteLines(tplEmpty.get(i)));
            }
            for (int i = tplEmpty.size(); i < conEmpty.size(); i++) {
                changes.addAll(insertLines(conEmpty.get(i)));
            }
            return changes;
        }

        if (tplSegments.size() == conSegments.size()
                && tplSegments.size() > 1
                && segmentsSameKind(tplSegments, conSegments)) {
            return diffBySegments(tplSegments, conSegments);
        }

        return diffLineByLine(templateLines, contractLines);
    }

    private boolean segmentsSameKind(List<Segment> tplSegments, List<Segment> conSegments) {
        for (int i = 0; i < tplSegments.size(); i++) {
            if (!tplSegments.get(i).kind().equals(conSegments.get(i).kind())) {
                return false;
            }
        }
        return true;
    }

    private List<Segment> splitSegments(List<LineUnit> lines) {
        List<Segment> segments = new ArrayList<>();
        int index = 0;

        while (index < lines.size()) {
            if (lines.get(index).normalized().isEmpty()) {
                int start = index;
                while (index < lines.size() && lines.get(index).normalized().isEmpty()) {
                    index++;
                }
                segments.add(new Segment("empty", lines.subList(start, index)));
            } else {
                int start = index;
                index++;
                while (index < lines.size()) {
                    if (lines.get(index).normalized().isEmpty()) {
                        break;
                    }
                    if (isParagraphBreak(lines.get(index - 1), lines.get(index))) {
                        break;
                    }
                    index++;
                }
                segments.add(new Segment("content", lines.subList(start, index)));
            }
        }
        return segments;
    }

    private boolean isParagraphBreak(LineUnit prev, LineUnit curr) {
        if (prev.page() != curr.page()) {
            return true;
        }
        double prevHeight = Math.max(prev.bbox()[3] - prev.bbox()[1], 1.0);
        double gap = curr.bbox()[1] - prev.bbox()[3];
        return gap > prevHeight * 0.75;
    }

    private List<RawChange> diffBySegments(List<Segment> tplSegments, List<Segment> conSegments) {
        List<RawChange> changes = new ArrayList<>();
        for (int i = 0; i < tplSegments.size(); i++) {
            List<LineUnit> tplChunk = tplSegments.get(i).lines();
            List<LineUnit> conChunk = conSegments.get(i).lines();
            if (!tplChunk.isEmpty() && !tplChunk.get(0).normalized().isEmpty()) {
                changes.addAll(diffContentSection(tplChunk, conChunk));
            } else {
                changes.addAll(diffEmptyRun(tplChunk, conChunk));
            }
        }
        return changes;
    }

    private List<RawChange> diffContentSection(List<LineUnit> tpl, List<LineUnit> con) {
        List<RawChange> changes = new ArrayList<>();
        int i = 0;
        int j = 0;

        while (i < tpl.size() && j < con.size()) {
            if (tpl.get(i).normalized().equals(con.get(j).normalized())) {
                i++;
                j++;
                continue;
            }

            Integer tplInCon = findLineIn(tpl.get(i).normalized(), con, j + 1, con.size());
            Integer conInTpl = findLineIn(con.get(j).normalized(), tpl, i + 1, tpl.size());

            if (tplInCon != null && conInTpl == null) {
                changes.add(RawChange.insert(con.get(j)));
                j++;
            } else if (conInTpl != null && tplInCon == null) {
                changes.add(RawChange.delete(tpl.get(i)));
                i++;
            } else if (tplInCon != null && conInTpl != null) {
                if (tplInCon - j <= conInTpl - i) {
                    changes.add(RawChange.insert(con.get(j)));
                    j++;
                } else {
                    changes.add(RawChange.delete(tpl.get(i)));
                    i++;
                }
            } else {
                changes.add(RawChange.delete(tpl.get(i)));
                i++;
            }
        }

        changes.addAll(deleteLines(tpl.subList(i, tpl.size())));
        changes.addAll(insertLines(con.subList(j, con.size())));
        return changes;
    }

    private List<RawChange> diffEmptyRun(List<LineUnit> tpl, List<LineUnit> con) {
        int pairCount = Math.min(tpl.size(), con.size());
        List<RawChange> changes = new ArrayList<>();
        if (pairCount < tpl.size()) {
            changes.addAll(deleteLines(tpl.subList(pairCount, tpl.size())));
        }
        if (pairCount < con.size()) {
            changes.addAll(insertLines(con.subList(pairCount, con.size())));
        }
        return changes;
    }

    private List<RawChange> deleteLines(List<LineUnit> lines) {
        List<RawChange> changes = new ArrayList<>();
        for (LineUnit line : lines) {
            changes.add(RawChange.delete(line));
        }
        return changes;
    }

    private List<RawChange> insertLines(List<LineUnit> lines) {
        List<RawChange> changes = new ArrayList<>();
        for (LineUnit line : lines) {
            changes.add(RawChange.insert(line));
        }
        return changes;
    }

    private List<RawChange> diffLineByLine(List<LineUnit> tpl, List<LineUnit> con) {
        return diffRange(tpl, con, 0, tpl.size(), 0, con.size());
    }

    private List<RawChange> diffRange(
            List<LineUnit> tpl,
            List<LineUnit> con,
            int i,
            int iEnd,
            int j,
            int jEnd
    ) {
        List<RawChange> changes = new ArrayList<>();

        while (i < iEnd && j < jEnd) {
            LineUnit tplLine = tpl.get(i);
            LineUnit conLine = con.get(j);

            if (tplLine.normalized().equals(conLine.normalized())) {
                i++;
                j++;
                continue;
            }

            if (tplLine.normalized().isEmpty() && !conLine.normalized().isEmpty()) {
                if (i + 1 < iEnd && tpl.get(i + 1).normalized().equals(conLine.normalized())) {
                    changes.add(RawChange.delete(tplLine));
                    i++;
                    continue;
                }
            } else if (!tplLine.normalized().isEmpty() && conLine.normalized().isEmpty()) {
                if (j + 1 < jEnd && con.get(j + 1).normalized().equals(tplLine.normalized())) {
                    changes.add(RawChange.insert(conLine));
                    j++;
                    continue;
                }
            }

            Integer tplInCon = findLineIn(tplLine.normalized(), con, j + 1, jEnd);
            Integer conInTpl = findLineIn(conLine.normalized(), tpl, i + 1, iEnd);

            if (tplInCon != null && conInTpl == null) {
                changes.add(RawChange.insert(conLine));
                j++;
            } else if (conInTpl != null && tplInCon == null) {
                changes.add(RawChange.delete(tplLine));
                i++;
            } else if (tplInCon != null && conInTpl != null) {
                if (tplInCon - j <= conInTpl - i) {
                    changes.add(RawChange.insert(conLine));
                    j++;
                } else {
                    changes.add(RawChange.delete(tplLine));
                    i++;
                }
            } else {
                changes.add(RawChange.delete(tplLine));
                i++;
            }
        }

        changes.addAll(deleteLines(tpl.subList(i, iEnd)));
        changes.addAll(insertLines(con.subList(j, jEnd)));
        return changes;
    }

    private Integer findLineIn(String text, List<LineUnit> lines, int start, int end) {
        for (int idx = start; idx < end; idx++) {
            if (lines.get(idx).normalized().equals(text)) {
                return idx;
            }
        }
        return null;
    }
}

File: backend-go/internal/service/diff_engine.go
package service

import "github.com/pdfdiff/backend-go/internal/domain"

func DiffLines(templateLines, contractLines []domain.LineUnit) []domain.RawChange {
	tplSegments := splitSegments(templateLines)
	conSegments := splitSegments(contractLines)

	tplContent := filterSegments(tplSegments, "content")
	conContent := filterSegments(conSegments, "content")
	tplEmpty := filterSegments(tplSegments, "empty")
	conEmpty := filterSegments(conSegments, "empty")

	if len(tplContent) == len(conContent) && len(tplContent) >= 2 {
		var changes []domain.RawChange
		for i := range tplContent {
			changes = append(changes, diffContentSection(tplContent[i], conContent[i])...)
		}
		emptyPairs := min(len(tplEmpty), len(conEmpty))
		for i := 0; i < emptyPairs; i++ {
			changes = append(changes, diffEmptyRun(tplEmpty[i], conEmpty[i])...)
		}
		for i := len(conEmpty); i < len(tplEmpty); i++ {
			changes = append(changes, deleteLines(tplEmpty[i])...)
		}
		for i := len(tplEmpty); i < len(conEmpty); i++ {
			changes = append(changes, insertLines(conEmpty[i])...)
		}
		return changes
	}

	if len(tplSegments) == len(conSegments) && len(tplSegments) > 1 && segmentsSameKind(tplSegments, conSegments) {
		return diffBySegments(tplSegments, conSegments)
	}

	return diffLineByLine(templateLines, contractLines)
}

type lineSlice []domain.LineUnit

func (s lineSlice) size() int { return len(s) }

func filterSegments(segments []domain.Segment, kind string) []lineSlice {
	var out []lineSlice
	for _, seg := range segments {
		if seg.Kind == kind {
			out = append(out, seg.Lines)
		}
	}
	return out
}

func segmentsSameKind(tpl, con []domain.Segment) bool {
	for i := range tpl {
		if tpl[i].Kind != con[i].Kind {
			return false
		}
	}
	return true
}

func splitSegments(lines []domain.LineUnit) []domain.Segment {
	var segments []domain.Segment
	index := 0

	for index < len(lines) {
		if lines[index].Normalized == "" {
			start := index
			for index < len(lines) && lines[index].Normalized == "" {
				index++
			}
			segments = append(segments, domain.Segment{Kind: "empty", Lines: lines[start:index]})
		} else {
			start := index
			index++
			for index < len(lines) {
				if lines[index].Normalized == "" {
					break
				}
				if isParagraphBreak(lines[index-1], lines[index]) {
					break
				}
				index++
			}
			segments = append(segments, domain.Segment{Kind: "content", Lines: lines[start:index]})
		}
	}
	return segments
}

func isParagraphBreak(prev, curr domain.LineUnit) bool {
	if prev.Page != curr.Page {
		return true
	}
	prevHeight := prev.BBox[3] - prev.BBox[1]
	if prevHeight < 1 {
		prevHeight = 1
	}
	gap := curr.BBox[1] - prev.BBox[3]
	return gap > prevHeight*0.75
}

func diffBySegments(tplSegments, conSegments []domain.Segment) []domain.RawChange {
	var changes []domain.RawChange
	for i := range tplSegments {
		tplChunk := tplSegments[i].Lines
		conChunk := conSegments[i].Lines
		if len(tplChunk) > 0 && tplChunk[0].Normalized != "" {
			changes = append(changes, diffContentSection(tplChunk, conChunk)...)
		} else {
			changes = append(changes, diffEmptyRun(tplChunk, conChunk)...)
		}
	}
	return changes
}

func diffContentSection(tpl, con []domain.LineUnit) []domain.RawChange {
	var changes []domain.RawChange
	i, j := 0, 0

	for i < len(tpl) && j < len(con) {
		if tpl[i].Normalized == con[j].Normalized {
			i++
			j++
			continue
		}

		tplInCon := findLineIn(tpl[i].Normalized, con, j+1, len(con))
		conInTpl := findLineIn(con[j].Normalized, tpl, i+1, len(tpl))

		switch {
		case tplInCon >= 0 && conInTpl < 0:
			changes = append(changes, domain.InsertChange(con[j]))
			j++
		case conInTpl >= 0 && tplInCon < 0:
			changes = append(changes, domain.DeleteChange(tpl[i]))
			i++
		case tplInCon >= 0 && conInTpl >= 0:
			if tplInCon-j <= conInTpl-i {
				changes = append(changes, domain.InsertChange(con[j]))
				j++
			} else {
				changes = append(changes, domain.DeleteChange(tpl[i]))
				i++
			}
		default:
			changes = append(changes, domain.DeleteChange(tpl[i]))
			i++
		}
	}

	changes = append(changes, deleteLines(tpl[i:])...)
	changes = append(changes, insertLines(con[j:])...)
	return changes
}

func diffEmptyRun(tpl, con []domain.LineUnit) []domain.RawChange {
	pairCount := min(len(tpl), len(con))
	var changes []domain.RawChange
	if pairCount < len(tpl) {
		changes = append(changes, deleteLines(tpl[pairCount:])...)
	}
	if pairCount < len(con) {
		changes = append(changes, insertLines(con[pairCount:])...)
	}
	return changes
}

func deleteLines(lines []domain.LineUnit) []domain.RawChange {
	changes := make([]domain.RawChange, 0, len(lines))
	for _, line := range lines {
		changes = append(changes, domain.DeleteChange(line))
	}
	return changes
}

func insertLines(lines []domain.LineUnit) []domain.RawChange {
	changes := make([]domain.RawChange, 0, len(lines))
	for _, line := range lines {
		changes = append(changes, domain.InsertChange(line))
	}
	return changes
}

func diffLineByLine(tpl, con []domain.LineUnit) []domain.RawChange {
	return diffRange(tpl, con, 0, len(tpl), 0, len(con))
}

func diffRange(tpl, con []domain.LineUnit, i, iEnd, j, jEnd int) []domain.RawChange {
	var changes []domain.RawChange

	for i < iEnd && j < jEnd {
		tplLine := tpl[i]
		conLine := con[j]

		if tplLine.Normalized == conLine.Normalized {
			i++
			j++
			continue
		}

		if tplLine.Normalized == "" && conLine.Normalized != "" {
			if i+1 < iEnd && tpl[i+1].Normalized == conLine.Normalized {
				changes = append(changes, domain.DeleteChange(tplLine))
				i++
				continue
			}
		} else if tplLine.Normalized != "" && conLine.Normalized == "" {
			if j+1 < jEnd && con[j+1].Normalized == tplLine.Normalized {
				changes = append(changes, domain.InsertChange(conLine))
				j++
				continue
			}
		}

		tplInCon := findLineIn(tplLine.Normalized, con, j+1, jEnd)
		conInTpl := findLineIn(conLine.Normalized, tpl, i+1, iEnd)

		switch {
		case tplInCon >= 0 && conInTpl < 0:
			changes = append(changes, domain.InsertChange(conLine))
			j++
		case conInTpl >= 0 && tplInCon < 0:
			changes = append(changes, domain.DeleteChange(tplLine))
			i++
		case tplInCon >= 0 && conInTpl >= 0:
			if tplInCon-j <= conInTpl-i {
				changes = append(changes, domain.InsertChange(conLine))
				j++
			} else {
				changes = append(changes, domain.DeleteChange(tplLine))
				i++
			}
		default:
			changes = append(changes, domain.DeleteChange(tplLine))
			i++
		}
	}

	changes = append(changes, deleteLines(tpl[i:iEnd])...)
	changes = append(changes, insertLines(con[j:jEnd])...)
	return changes
}

func findLineIn(text string, lines []domain.LineUnit, start, end int) int {
	for idx := start; idx < end; idx++ {
		if lines[idx].Normalized == text {
			return idx
		}
	}
	return -1
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
